### PR TITLE
Deck.gl: basic support for polygons and WMS layers

### PR DESCRIPTION
**Pull request by vlach1989**
See ptr-docs for examples:

/components/maps/presentational/deckGl
/architecture/storeDataTypes/common/style/vectors

### CODE REVIEW

File: src/DeckGlMap/helpers/style.js [hunk 1]
```diff
@@ -130,36 +130,31 @@ function getRgbaColorArray(rgbColorArray, opacity) {
...
-			return getStyleObjectForAttribute(
-				style.attributeStyles[0],
-				style.baseStyle,
-				feature.properties
-			);
-		} else {
-			// TODO merge styles if there are styles for more attributes
-		}
+	if (style.attributeStyles?.length) {
+		let finalStyle = {...style.baseStyle};
```
Comment (vdubr) [inline src/DeckGlMap/helpers/style.js:134]: You can use const instead of let

File: src/DeckGlMap/layers/VectorLayer.js [hunk 2]
```diff
@@ -1,11 +1,40 @@
...
+import {
+	forIn as _forIn,
+	includes as _includes,
+	partition as _partition,
+} from 'lodash';
 import styleHelpers from '../helpers/style';
+import constants from '../../constants';
 
 class VectorLayer extends CompositeLayer {
+	constructor(props) {
```
Comment (vdubr) [inline src/DeckGlMap/layers/VectorLayer.js:12]: This is not necessary if you don't declare any property.

File: src/DeckGlMap/layers/VectorLayer.js [hunk 3]
```diff
@@ -126,6 +165,26 @@ class VectorLayer extends CompositeLayer {
 		return defaultStyle?.outlineWidth;
 	}
 
+	/**
+	 * Return all selected feature keys
+	 * @param selections {Object}
+	 * @returns {Array|null}
+	 */
+	getSelectedFeatureKeys(selections) {
+		if (selections) {
```
Comment (vdubr) [inline src/DeckGlMap/layers/VectorLayer.js:174]:
```suggestion
		if (selections?.length > 0) {
```

File: src/DeckGlMap/layers/WmsLayer.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
+import {CompositeLayer} from '@deck.gl/core';
+import {TileLayer} from '@deck.gl/geo-layers';
+import {BitmapLayer} from '@deck.gl/layers';
+import {load} from '@loaders.gl/core';
+import SphericalMercator from '@mapbox/sphericalmercator';
+
+const conversion = new SphericalMercator({size: 512, antimeridian: true});
+
+class WmsLayer extends CompositeLayer {
```
Comment (vdubr) [inline src/DeckGlMap/layers/WmsLayer.js:9]: Is there some reason, why use "class" and not arow?

File: src/DeckGlMap/layers/WmsLayer.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
+import {CompositeLayer} from '@deck.gl/core';
+import {TileLayer} from '@deck.gl/geo-layers';
+import {BitmapLayer} from '@deck.gl/layers';
+import {load} from '@loaders.gl/core';
+import SphericalMercator from '@mapbox/sphericalmercator';
+
+const conversion = new SphericalMercator({size: 512, antimeridian: true});
```
Comment (vdubr) [inline src/DeckGlMap/layers/WmsLayer.js:7]: size=512 means that tiles have size 512X512 px? Standard is 256px and 512px are used usualy on hiDPI devices.

File: src/DeckGlMap/layers/WmsLayer.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	renderWmsLayer() {
+		const {options, opacity, key} = this.props;
+		if (!options) {
+			throw new Error('WmsLayer: options are not defined!');
+		}
+		const {url, params} = options;
+		if (!url) {
+			throw new Error('WmsLayer: options.url is not defined!');
+		}
+		if (!url) {
```
Comment (vdubr) [inline src/DeckGlMap/layers/WmsLayer.js:23]:
```suggestion
		if (!params) {
```

File: src/DeckGlMap/layers/WmsLayer.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+			getTileData: tile => {
+				const {x, y, z} = tile;
+
+				// Conversion needed due to visualizations issues.
+				const [west, south, east, north] = conversion.bbox(
+					x,
+					y,
+					z,
+					false,
+					'900913'
```
Comment (vdubr) [inline src/DeckGlMap/layers/WmsLayer.js:45]: 😢  I did not used this projection so long...

File: src/DeckGlMap/layers/WmsLayer.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+		const id = `${key}-wmsLayer`;
+
+		return new TileLayer({
+			id,
+			opacity,
+			getTileData: tile => {
+				const {x, y, z} = tile;
+
+				// Conversion needed due to visualizations issues.
+				const [west, south, east, north] = conversion.bbox(
```
Comment (vdubr) [inline src/DeckGlMap/layers/WmsLayer.js:40]: Probably tile size [512] should be same like "tileSize". Did you tested WMS with 256 and 512 tileSize?

File: src/DeckGlMap/layers/WmsLayer.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+					x,
+					y,
+					z,
+					false,
+					'900913'
+				);
+
+				let urlQueryStringParams = {
```
Comment (vdubr) [inline src/DeckGlMap/layers/WmsLayer.js:48]:
```suggestion
				const urlQueryStringParams = {
```

File: src/DeckGlMap/layers/WmsLayer.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+					bbox: [west, south, east, north].join(','),
+					format: format || 'image/png',
+					height: tileSize || 512,
+					layers,
+					request: 'GetMap',
+					service: 'WMS',
+					styles: styles || '',
+					width: tileSize || 512,
+					transparent: true,
```
Comment (vdubr) [inline src/DeckGlMap/layers/WmsLayer.js:57]:
```suggestion
					transparent: params?.transparent === true ? true : false,
```

File: src/DeckGlMap/layers/WmsLayer.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+				};
+
+				if (params.hasOwnProperty('transparent')) {
+					urlQueryStringParams.transparent = params.transparent;
+				}
+
+				if (version === '1.3.0') {
```
Comment (vdubr) [inline src/DeckGlMap/layers/WmsLayer.js:64]: It means that version 1.1.1 is default. I would prefer 1.3.0 as a default.

File: src/DeckGlMap/layers/WmsLayer.js (hunk 4 above)
Comment (vlach1989) [reply src/DeckGlMap/layers/WmsLayer.js:9]: Yes, it's recommended https://deck.gl/docs/developer-guide/custom-layers/composite-layers

File: src/DeckGlMap/layers/WmsLayer.js (hunk 7 above)
Comment (vlach1989) [reply src/DeckGlMap/layers/WmsLayer.js:45]: See: convert method here https://www.npmjs.com/package/@mapbox/sphericalmercator

File: src/DeckGlMap/layers/WmsLayer.js (hunk 10 above)
Comment (vlach1989) [reply src/DeckGlMap/layers/WmsLayer.js:57]: From my POV it's better if true is the default value

File: src/DeckGlMap/layers/VectorLayer.js (hunk 3 above)
Comment (vlach1989) [reply src/DeckGlMap/layers/VectorLayer.js]: https://github.com/gisat-panther/ptr-maps/pull/116/commits/45ba7e5e118cd62d9c70e3f5197ee133af9c00b4

File: src/DeckGlMap/layers/VectorLayer.js (hunk 2 above)
Comment (vlach1989) [reply src/DeckGlMap/layers/VectorLayer.js]: https://github.com/gisat-panther/ptr-maps/pull/116/commits/257b71898db9b9366ccf0c41dd5007b504abd525

File: src/DeckGlMap/layers/WmsLayer.js (hunk 11 above)
Comment (vlach1989) [reply src/DeckGlMap/layers/WmsLayer.js]: https://github.com/gisat-panther/ptr-maps/pull/116/commits/4c411e12e9dbd3f56f98127289559f569561c3ab

File: src/DeckGlMap/layers/WmsLayer.js (hunk 8 above)
Comment (vlach1989) [reply src/DeckGlMap/layers/WmsLayer.js:47]: https://github.com/gisat-panther/ptr-maps/pull/116/commits/4c411e12e9dbd3f56f98127289559f569561c3ab

File: src/DeckGlMap/layers/WmsLayer.js (hunk 5 above)
Comment (vlach1989) [reply src/DeckGlMap/layers/WmsLayer.js]: https://github.com/gisat-panther/ptr-maps/pull/116/commits/4c411e12e9dbd3f56f98127289559f569561c3ab

**Review by vdubr — APPROVED** [review]

**Pull request merged**